src/sync/all_forks.rs

```diff
@@ -976,13 +883,153 @@ impl<TBl, TRq, TSrc> ops::IndexMut<SourceId> for AllForksSync<TBl, TRq, TSrc> {
     }
 }
 
-/// Struct to pass back when a block request has succeeded.
-#[derive(Debug)]
-pub struct RequestSuccessBlock<THdr, TJs> {
-    /// SCALE-encoded header returned by the remote.
-    pub scale_encoded_header: THdr,
-    /// SCALE-encoded justification returned by the remote.
-    pub scale_encoded_justifications: TJs,
+/// See [`AllForksSync::finish_ancestry_search`].
+pub struct FinishAncestrySearch<TBl, TRq, TSrc> {
+    inner: AllForksSync<TBl, TRq, TSrc>,
+
+    /// Source that has sent the request that is being answered.
+    source_id: SourceId,
+
+    /// Set to true if any block at all have been added.
+    any_progress: bool,
+
+    /// Number of blocks added before through that data structure.
+    index_in_response: usize,
+
+    /// Hash of the block that was initially request.
+    requested_block_hash: [u8; 32],
+    /// Height of the block that was initially request.
+    requested_block_height: u64,
+
+    /// The next block to add should have a hash equal to this one.
+    expected_next_hash: [u8; 32],
+    /// The next block to add should have a height equal to this one.
+    expected_next_height: u64,
+}
+
+impl<TBl, TRq, TSrc> FinishAncestrySearch<TBl, TRq, TSrc> {
+    /// Adds a block coming from the response that the source has provided.
+    ///
+    /// On success, the block is inserted in the state machine.
+    ///
+    /// If an error is returned, the [`FinishAncestrySearch`] is turned back again into a
+    /// [`AllForksSync`], but all the blocks that have already been added are retained.
+    pub fn add_block(
+        mut self,
+        scale_encoded_header: impl AsRef<[u8]>,
+        scale_encoded_justifications: impl Iterator<Item = ([u8; 4], impl AsRef<[u8]>)>,
+    ) -> Result<Self, (AncestrySearchResponseError, AllForksSync<TBl, TRq, TSrc>)> {
+        let scale_encoded_header = scale_encoded_header.as_ref();
```

Comment:
isn't `scale_encoded_header` already a ref?

Reply:
`impl AsRef<[u8]>` allows you pass things such as `Vec<u8>` and such. It doesn't strictly have to be a `&[u8]`.
I've originally used `impl AsRef<[u8]>` because when we're talking about the items yielded by an iterator it's sometimes the only solution. For example if you have an iterator that generates `Vec<u8>`s, you can't turn it into an iterator that generates `&[u8]`s without some additional buffering.

However, now that the iterator is gone, I'm going to turn this into a `&[u8]`.